src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/IconUrlToImageCacheConverter.cs

```diff
@@ -24,38 +30,114 @@ internal class IconUrlToImageCacheConverter : IValueConverter
             WebExceptionStatus.UnknownError
         };
 
-        private static readonly System.Net.Cache.RequestCachePolicy RequestCacheIfAvailable = new System.Net.Cache.RequestCachePolicy(System.Net.Cache.RequestCacheLevel.CacheIfAvailable);
+        private static readonly RequestCachePolicy RequestCacheIfAvailable = new RequestCachePolicy(RequestCacheLevel.CacheIfAvailable);
 
         private static readonly ErrorFloodGate _errorFloodGate = new ErrorFloodGate();
 
+#if DEBUG
+        public StringBuilder Messages { get; set; }
+
+        private void LogMessage(string message)
+        {
+            Messages?.AppendFormat("IconUrlToImageCacheConverter: {0}{1}", message, Environment.NewLine);
+        }
+#endif
+
         // We bind to a BitmapImage instead of a Uri so that we can control the decode size, since we are displaying 32x32 images, while many of the images are 128x128 or larger.
         // This leads to a memory savings.
         public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
         {
+#if DEBUG
+            LogMessage($"Convert Params: {value.ToString()}, {targetType.ToString()}, {parameter.ToString()}, {culture.ToString()}");
+#endif
             var iconUrl = value as Uri;
             var defaultPackageIcon = parameter as BitmapSource;
             if (iconUrl == null)
             {
+#if DEBUG
+                LogMessage($"Convert iconUrl is null");
+#endif
                 return null;
             }
 
             var cachedBitmapImage = _bitmapImageCache.Get(iconUrl.ToString()) as BitmapSource;
             if (cachedBitmapImage != null)
             {
+#if DEBUG
+                LogMessage($"Convert Cache Hit: {cachedBitmapImage.ToString()}");
+#endif
                 return cachedBitmapImage;
             }
 
             // Some people run on networks with internal NuGet feeds, but no access to the package images on the internet.
             // This is meant to detect that kind of case, and stop spamming the network, so the app remains responsive.
             if (_errorFloodGate.IsOpen)
             {
+#if DEBUG
+                LogMessage($"Convert errorFloodGate open");
+#endif
                 return defaultPackageIcon;
             }
 
             var iconBitmapImage = new BitmapImage();
             iconBitmapImage.BeginInit();
-            iconBitmapImage.UriSource = iconUrl;
 
+            var markIdx = iconUrl.OriginalString.LastIndexOf("#");
+
+            BitmapSource imageResult = null;
+
+            if (iconUrl.IsAbsoluteUri && iconUrl.IsFile && markIdx >= 0)
+            {
+#if DEBUG
+                LogMessage("Convert EmbeddedIcon");
+#endif
+                try
+                {
+                    using (var par = new PackageArchiveReader(Uri.UnescapeDataString(iconUrl.LocalPath)))
```

Comment:
With this change, for every local package will load the complete archive in memory twice.
Once in the LocalPackageSearchMetadata -> LocalPackageInfo, PackageItemListViewModel, and then again in the icon to url converter. 

We should try to make this cheaper.

Reply:
Some ideas:
1. I can try passing a PackageArchiveReader instance as a parameter to the IconUrlToImageCacheConverter
2. I can convert IconUrlToImageCacheConverter to a MultiValueConverter, and accept any number of parameters as needed

Reply:
A separate PR is available with the idea of recycling the underlying LocalPackageInfo to get a PackageArchiveReader (PAR). The problem with that approcach is that the function that generates PARs always create a new object on each call.

https://github.com/NuGet/NuGet.Client/pull/3153